cmd/neofs-cli/modules/tree/add.go

```diff
@@ -0,0 +1,119 @@
+package tree
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/nspcc-dev/neofs-node/cmd/neofs-cli/internal/common"
+	"github.com/nspcc-dev/neofs-node/cmd/neofs-cli/internal/commonflags"
+	"github.com/nspcc-dev/neofs-node/cmd/neofs-cli/internal/key"
+	"github.com/nspcc-dev/neofs-node/pkg/services/tree"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"github.com/spf13/cobra"
+)
+
+var addCmd = &cobra.Command{
+	Use:   "add",
+	Short: "Add a node to the tree service",
+	Run:   add,
+	PersistentPreRun: func(cmd *cobra.Command, _ []string) {
+		commonflags.Bind(cmd)
+	},
+}
+
+func initAddCmd() {
+	commonflags.Init(addCmd)
+	initCTID(addCmd)
+
+	ff := addCmd.Flags()
+	ff.StringSlice(metaFlagKey, nil, "Meta pairs in the form of Key1=[0x]Value1,Key2=[0x]Value2")
+	ff.Uint64(pidFlagKey, 0, "Parent node ID")
+
+	_ = cobra.MarkFlagRequired(ff, commonflags.RPC)
+}
+
+func add(cmd *cobra.Command, _ []string) {
+	pk := key.GetOrGenerate(cmd)
+
+	var cnr cid.ID
+	err := cnr.DecodeString(cmd.Flag(cidFlagKey).Value.String())
+	common.ExitOnErr(cmd, "decode container ID string: %w", err)
+
+	tid := cmd.Flag(tidFlagKey).Value.String()
+	if tid == "" {
+		common.ExitOnErr(cmd, "", errors.New("empty tree ID"))
+	}
+
+	pid, _ := cmd.Flags().GetUint64(pidFlagKey)
+
+	meta, err := parseMeta(cmd)
+	common.ExitOnErr(cmd, "meta data parsing: %w", err)
+
+	ctx := cmd.Context()
+
+	cli, err := _client(ctx)
+	common.ExitOnErr(cmd, "client: %w", err)
+
+	rawCID := make([]byte, sha256.Size)
+	cnr.Encode(rawCID)
+
+	req := new(tree.AddRequest)
+	req.Body = &tree.AddRequest_Body{
+		ContainerId: rawCID,
+		TreeId:      tid,
+		ParentId:    pid,
+		Meta:        meta,
+		BearerToken: nil, // TODO: add token handling
+	}
+
+	common.ExitOnErr(cmd, "message signing: %w", tree.SignMessage(req, pk))
+
+	resp, err := cli.Add(ctx, req)
+	common.ExitOnErr(cmd, "rpc call: %w", err)
+
+	cmd.Println("Node ID: ", resp.Body.NodeId)
+}
+
+func parseMeta(cmd *cobra.Command) ([]*tree.KeyValue, error) {
+	raws, _ := cmd.Flags().GetStringSlice(metaFlagKey)
+	if len(raws) == 0 {
+		return nil, nil
+	}
+
+	var err error
+
+	pairs := make([]*tree.KeyValue, 0, len(raws))
+	for i := range raws {
+		kv := strings.SplitN(raws[i], "=", 2)
+		if len(kv) != 2 {
+			return nil, fmt.Errorf("invalid meta pair format: %s", raws[i])
+		}
+
+		var pair tree.KeyValue
+		pair.Key = kv[0]
+		pair.Value, err = bytesFromString(kv[1])
+		if err != nil {
+			return nil, fmt.Errorf("invalid meta value: %w", err)
+		}
+
+		pairs = append(pairs, &pair)
+	}
+
+	return pairs, nil
+}
+
+func bytesFromString(s string) ([]byte, error) {
+	switch {
+	case strings.HasPrefix(s, "0x"):
```

Comment:
Do we need this in CLI? Especially when we don't have similar logic in get operations now.

Reply:
> Do we need this in CLI

i do not have info about how values in the Tree's KV gonna (planned to) be used, so could not answer that question at all

> Especially when we don't have similar logic in get operations now.

i am ready to add that but dont know how that is expected to work (how to distinguish string bytes from the raw bytes)

in a word -- dropped